src/java.base/share/classes/jdk/crac/impl/OpenFDPolicies.java

```diff
@@ -0,0 +1,307 @@
+                } else {
+                    properties.put(unescape(item, 0, eqIndex), item.substring(eqIndex + 1));
+                }
+            }
+        }
+        return properties;
+    }
+
+    private static String unescape(String str, int fromIndex, int toIndex) {
+        boolean escaped = false;
+        StringBuilder sb = new StringBuilder(str.length() - fromIndex);
+        for (int i = fromIndex; i < toIndex; ++i) {
+            char c = str.charAt(i);
+            if (!escaped && c == '\\') {
+                escaped = true;
+            } else {
+                sb.append(c);
+                escaped = false;
+            }
+        }
+        return sb.toString();
+    }
+
+    private static int findNonEscapedEq(String str, int fromIndex) {
+        boolean escaped = false;
+        for (int i = fromIndex; i < str.length(); ++i) {
+            char c = str.charAt(i);
+            if (c == '\\') {
+                escaped = !escaped;
+            } else if (c == '=' && !escaped) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    public static void ensureRegistered() {
+        // Noop - this method is invoked to ensure that static constructor was invoked
+    }
+
+    private void setPolicy(Object key, Object value, P policy) {
+        if (FIFO.equals(key)) {
+            fifoPolicy = policy;
+        } else if (SOCKET.equals(key)) {
+            socketPolicy = policy;
+        } else if (NUMERIC.matcher(key.toString()).matches()) {
+            int fd = Integer.parseInt(key.toString());
+            P prev = numericPolicies.putIfAbsent(fd, policy);
+            if (prev != null) {
+                LoggerContainer.error("Duplicate policy for file descriptor {0}; policy {1} will be ignored.", fd, value);
+            }
+        } else {
+            pathPolicies.add(Map.entry(
+                    FileSystems.getDefault().getPathMatcher("glob:" + key), policy));
+        }
+    }
+
+    public P get(int fd, String type, String path) {
+        synchronized (OpenFDPolicies.this) {
+            if (!loaded) {
+                // We could use a Runnable but method references don't work
+                // when the static ctor is invoked
+                if (this == CHECKPOINT) {
+                    loadCheckpointPolicies();
+                } else {
+                    loadRestorePolicies();
+                }
+            }
+        }
+        P policy = numericPolicies.get(fd);
+        if (policy != null) {
+            return policy;
+        }
+        if (path != null) {
+            Path p = Path.of(path);
+            for (var entry : pathPolicies) {
+                if (entry.getKey().matches(p)) {
+                    return entry.getValue();
+                }
+            }
+        }
+        if (type.equals("fifo") && fd > 2) {
+            return fifoPolicy;
+        } else if (type.equals("socket")) {
+            return socketPolicy;
```

Comment:
Obviously, only a single policy is possible for all sockets. In general sockets are much harder than files, and in not so many cases we can automatically handle them. I think some connection-less sockets _may_ be covered, and some listening sockets.

If we continue this implementation, FileDescriptor (a rather simple object intially) will grow larger and larger, knowing about all possible uses in the JDK.

Reply:
Someone already asked for automatic reopening of listening sockets. The policy could then e.g. filter specific ports (similar to files). Yes, you can't to anything but KEEP_CLOSED for sockets related to connection. I assume that the restore policy will be set during restore, rather than during checkpoint, so there's no chance to handle a non-sense configuration during checkpoint (because the configuration is not present yet).

Growing FileDescriptor has a point, any complex implementation for sockets should be hosted next to other Socket related stuff. For files, I don't know if there's a better place, but we could refactor it to an util class.

Reply:
> Someone already asked for automatic reopening of listening sockets.

It was me :) 

> The policy could then e.g. filter specific ports (similar to files). Yes, you can't to anything but KEEP_CLOSED for sockets related to connection.

Connection-less sockets may be fine as well. Maybe even stream sockets, with a big disclaimer that the policy administrator is fully responsible for breaking communication logic. By default, we should do the most safest thing. 

> I assume that the restore policy will be set during restore, rather than during checkpoint, so there's no chance to handle a non-sense configuration during checkpoint (because the configuration is not present yet).

I think this can be implemented https://github.com/openjdk/crac/pull/69#discussion_r1224461715

> Growing FileDescriptor has a point, any complex implementation for sockets should be hosted next to other Socket related stuff. For files, I don't know if there's a better place, but we could refactor it to an util class.

#79 does that with JDKFIleResource, which provides a limited hard-coded poilcy for files: exceptions are thrown only for files not in classpath.